xgi/utils/utilities.py

```diff
@@ -35,3 +36,42 @@ def dual_dict(edge_dict):
             node_dict[node].append(edge_id)
 
     return dict(node_dict)
+
+
+def powerset(iterable, include_empty=False, include_full=False):
+    """Returns all possible subsets of the elements in iterable, with options
+    to include the empty set and the set containing all elements.
+
+    Parameters
+    ----------
+    iterable : list-like
+        List of elements
+    include_empty: bool, default: False
+        Whether to include the empty set
+    include_full: bool, default: False
+        Whether to include the set containing all elements of iterable
+
+    Returns
+    -------
+    itertools.chain
+
+    Examples
+    --------
+    >>> list(powerset([1,2,3,4])) # doctest: +NORMALIZE_WHITESPACE
+    [(1,), (2,), (3,), (4,), (1, 2), (1, 3), (1, 4), (2, 3), (2, 4), (3, 4),
+     (1, 2, 3), (1, 2, 4), (1, 3, 4), (2, 3, 4)]
+
+    """
+
+    if include_empty:
+        start = 0
+    else:
+        start = 1
+
+    if include_full:
+        end = 1
+    else:
+        end = 0
+
+    s = list(iterable)
```

Comment:
Question: doesn't `combinations` accept an iterable? Why convert to list?

Reply:
Not sure, took it from itertools recipes (https://docs.python.org/3/library/itertools.html#itertools-recipes).
I'd keep it for safety, but shouldn't change much as the input should be a single edge.